File: strategy-pattern/go/src/context.go
package src

// 定义策略执行对象context
type Context struct {
	strategy Strategy
}

// 设置不同strategy，方法名首字母大写
func (c *Context) SetStrategy(s Strategy) {
	c.strategy = s
}

// 执行策略接口里面的方法
func (c *Context) Run() {
	c.strategy.Run()
}

File: strategy-pattern/c/src/context.c
#include "func.h"

// 定义策略执行对象context

// 设置不同strategy，方法名首字母大写
void context_set_strategy(Context *context, Strategy *strategy)
{
  context->strategy = strategy;
}

void context_run(Context *context)
{
  printf("\r\n context_run() [context=%s strategy=%s]", context->name, context->strategy->name);
  if (context->strategy != NULL)
  {
    context->strategy->run(context->strategy);
  }
}

Context *context_constructor(char *name)
{
  Context *context = (Context *)malloc(sizeof(Context));
  strcpy(context->name, name);
  context->set_strategy = &context_set_strategy;
  context->run = &context_run;
  return context;
}